add rule and loader to get generated tags

vonagam opened the pull request:
For #156

brunocodutra commented on src/index.js:
```diff
@@ -14,6 +15,24 @@ module.exports = class WebappWebpackPlugin {
+    assert(path.isAbsolute(this.options.logo), '`logo` must be an absolute path');
+
+    const rule = {
+      include: this.options.logo,
```
[path.isAbsolute](https://nodejs.org/api/path.html#path_path_isabsolute_path) accepts paths like `'/foo/../bar/logo.svg'`, but I'm not sure `rule.include` would match it against `'/bar/logo.svg'` correctly. 

Could you use `path.resolve` here to avoid issues?

vonagam replied:
Done

brunocodutra commented on src/ruleLoader.js:
```diff
@@ -0,0 +1,10 @@
+      callback(null, "module.exports = ['" + tags.join( "', '" ) + "'];");
+    })
+    .catch(callback);
+};
```
Would you mind renaming this to `rule_loader.js` to avoid issues with case insensitive file systems?

vonagam replied:
Done

brunocodutra commented on src/ruleLoader.js:
```diff
@@ -0,0 +1,10 @@
+
+  plugin.tags.promise
+    .then(tags => {
+      callback(null, "module.exports = ['" + tags.join( "', '" ) + "'];");
```
Could you use string templates instead?

`module.exports = ['${tags.join("', '")}'];`

vonagam replied:
Done

brunocodutra commented on test/rule.test.js:
```diff
@@ -0,0 +1,60 @@
+      path: dist,
+    },
+    plugins: [
+      new webpack.optimize.OccurrenceOrderPlugin(),
```
I'm not familiar with this plugin - what's the motivation?

vonagam replied:
Reproducible webpack builds.
Without this plugin running build twice on same machine will result in different js.
But it seems it does not help much as tests failing... (I am on mac)

brunocodutra replied:
When you say _reproducible_ what exactly can vary in each build?

brunocodutra commented:
Any idea why tests are [failing on Windows](https://ci.appveyor.com/project/brunocodutra/webapp-webpack-plugin/builds/21753346)?

vonagam commented:
The problem lies in non reproducible builds of js bundle.

To get reproducible builds on my personal machine (mac) i used [`OccurrenceOrderPlugin`](https://github.com/webpack/webpack/issues/739), seems it does not help with builds between machines.

brunocodutra commented:
What if we teach [`compare`](https://github.com/brunocodutra/webapp-webpack-plugin/blob/master/test/util.js#L61) to `eval`uate `main.js` (or any file that ends in `.js` for that matter), similarly to what we do [here](https://github.com/brunocodutra/webapp-webpack-plugin/blob/master/src/compiler.js#L49), and compare the output rather than the contents directly?

vonagam commented:
Added some description to README.

Fixed rule tests. Made bundle js output to export tags with cjs so that we can compare them with expected result by simply requiring them.

Offtop:
In cache tests you use [`t.pass(some true value)`](https://github.com/brunocodutra/webapp-webpack-plugin/blob/e583b5ac56042baa191616b9fe44c0b1ba913b2d/test/cache.custom.test.js#L22). I think you wanted to use `t.true` (or `t.truthy` for `length` value), because [`t.pass`](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#passmessage) accepts only message to show, so even if you call `t.pass(false)` it passes.

brunocodutra commented on README.md:
````diff
@@ -226,6 +226,34 @@ Example implementation:
   }
 ```
 
+## Rule
````
Although technically we create a _rule_ I think most users would look for this feature with the word _loader_ in mind, so what do you think about renaming this section to _Webpack Loader_ or something of that sort?

Then again I trust you to come up with the better term, after all you came up with the use case.

vonagam replied:
Hm... loader is good too. But people will not use loader directly (it is kinda internal implementation), what they receive and part of public api is getting rule to put into `module.rules`.

I do not care much about README description though. So if you think that you want something different feel free to request change or just modify it after merging.

brunocodutra replied:
Fair enough

brunocodutra commented on README.md:
```diff
@@ -226,6 +226,34 @@ Example implementation:
 
+## Rule
+
+To get generated html tags from js use plugin instance method `rule` by adding it's result to `module.rules`.
```
Typo `it's -> its`.

Generally I find this sentence hard to understand though, I suggest we go straight to the point and let the example speak for itself. 

I'd write something like

> A Webpack loader rule is also provided to make the generated HTML tags available to your JS app.

vonagam replied:
Done. (Replaced with suggestion)

brunocodutra commented on test/rule.test.js:
```diff
@@ -0,0 +1,65 @@
+  t.true(actualTags.length > 0);
+
+  const expectedTags = require('./fixtures/expected/rule/main');
+  t.deepEqual(actualTags, expectedTags);
```
Brilliant! There was really no point in checking the files on disk for this test, all we care about is that tags are injected into the bundle.

brunocodutra commented on src/index.js:
```diff
@@ -58,7 +79,10 @@ module.exports = class WebappWebpackPlugin {
         })
-        .catch(callback)
+        .catch(err => {
+          this.tags.reject(err);
```
I believe you need to add `this.tags.promise.catch(() => { /* ignore */ });` before this line to avoid the "Unhandled rejection errors" we're getting on CI.

brunocodutra replied:
@vonagam this seems to be the last blocker, once you address this issue and CI is green I'll merge this PR.

Thanks a lot for your hard work so far!

vonagam replied:
Done. (Added in constructor)

brunocodutra commented on src/index.js:
```diff
@@ -14,6 +15,25 @@ module.exports = class WebappWebpackPlugin {
+      this.tags.resolve = resolve;
+      this.tags.reject = reject;
+    });
+    this.tags.promise.catch(() => {});
```
Are you sure we can place the `catch` here?

I believe it would swallow the rejection from the plugin and the loader `then` would execute with `undefined`. So we need the `catch` to come after the rule is created, to let the loader catch the rejection and report the error.

Maybe we need a test case for that.

vonagam replied:
If i did it like this:
```js
this.tags.promise = new Promise((resolve, reject) => {
  this.tags.resolve = resolve;
  this.tags.reject = reject;
}).catch(() => {});
```
Then yea, rejection will be shallowed and rule loader will get `undefined` in `then`.

Here though, because original promise is exposed (and not `catch`ed), both `catch`es are called on a rejection. (Checked it before.)

vonagam replied:
Here is the snippet to check in dev tools:
```js
class Plugin {
  constructor() {
    this.tags = {};
    this.tags.promise = new Promise((resolve, reject) => {
      this.tags.resolve = resolve;
      this.tags.reject = reject;
    });
    this.tags.promise.catch(() => {
      console.log('original catch');
    });
  }
}

const plugin = new Plugin();

plugin.tags.promise
  .then(() => console.log('new then'))
  .catch(() => console.log('new catch'));

plugin.tags.reject('error');
```

brunocodutra replied:
Oh I see, I didn't know one could _fork_ promisses into multiple continuations.
